packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -140,89 +301,38 @@ export default createRule<Options, MessageIds>({
-      const discriminantType = getConstrainedTypeAtLocation(
-        services,
-        node.discriminant,
-      );
-      const symbolName = discriminantType.getSymbol()?.escapedName;
-
-      if (discriminantType.isUnion()) {
-        const unionTypes = tsutils.unionTypeParts(discriminantType);
-        const caseTypes = new Set<ts.Type>();
-        for (const switchCase of node.cases) {
-          if (switchCase.test == null) {
-            // Switch has 'default' branch - do nothing.
-            return;
-          }
-
-          caseTypes.add(
-            getConstrainedTypeAtLocation(services, switchCase.test),
-          );
-        }
-
-        const missingBranchTypes = unionTypes.filter(
-          unionType => !caseTypes.has(unionType),
-        );
+    function checkSwitchDangerousDefaultCase(
+      switchStatementMetadata: SwitchStatementMetadata,
+    ): void {
+      // This feature of the rule is gated behind an option.
+      if (allowDefaultCase) {
+        return;
+      }
 
-        if (missingBranchTypes.length === 0) {
-          // All cases matched - do nothing.
-          return;
-        }
+      const { missingBranchTypes, defaultCase } = switchStatementMetadata;
 
+      if (missingBranchTypes.length === 0 && defaultCase !== undefined) {
         context.report({
-          node: node.discriminant,
-          messageId: 'switchIsNotExhaustive',
-          data: {
-            missingBranches: missingBranchTypes
-              .map(missingType =>
-                tsutils.isTypeFlagSet(missingType, ts.TypeFlags.ESSymbolLike)
-                  ? `typeof ${missingType.getSymbol()?.escapedName as string}`
-                  : checker.typeToString(missingType),
-              )
-              .join(' | '),
-          },
-          suggest: [
-            {
-              messageId: 'addMissingCases',
-              fix(fixer): TSESLint.RuleFix {
-                return fixSwitch(
-                  fixer,
-                  node,
-                  missingBranchTypes,
-                  symbolName?.toString(),
-                );
-              },
-            },
-          ],
+          node: defaultCase,
+          messageId: 'dangerousDefaultCase',
         });
-      } else if (requireDefaultForNonUnion) {
-        const hasDefault = node.cases.some(
-          switchCase => switchCase.test == null,
-        );
-
-        if (!hasDefault) {
-          context.report({
-            node: node.discriminant,
-            messageId: 'switchIsNotExhaustive',
-            data: {
-              missingBranches: 'default',
-            },
-            suggest: [
-              {
-                messageId: 'addMissingCases',
-                fix(fixer): TSESLint.RuleFix {
-                  return fixSwitch(fixer, node, [null]);
-                },
-              },
-            ],
-          });
-        }
       }
     }
 
     return {
-      SwitchStatement: checkSwitchExhaustive,
```

Comment:
[Praise] I like how you split this up into functions 🙂 much easier to read through the main logic flow than keeping it all inline would be. Nice!